### ISSUE TITLE
unknown layer type Mean after using cv::dnn::Importer to init cv::dnn::Net

### ISSUE THREAD

**Issue by hhuangfh** [opencv/opencv#9661]
##### Environment
- OpenCV = 3.3
- Operating System / Platform => Win10 64 Bit
- Compiler => Visual Studio 2015
#####
Hi, everyone!
I tried to use  the dnn' API in OpenCV3.3 to extract image's feature, like the sample code "tf_inception.cpp". the model file is generated in the context of theOS include python, tensorflow, faceNet and so so and so.
At the beginning, there is an error shown like "Unknown layer type FIFOQueueV2 in op batch_join/fifo_queue)
and I use the tool called "optimize_for_inference" under "tensorflow/python/tools" to process the .pb model file.
run the command in cmd terminal, like:
--------------------------------------------------------------------
python optimize_for_inference .py  --input=frozen_graph.pb  --output=output_graph.pb  --input_names="input"  --output_names="embedding"

--------------------------------------------------------------------
But this time, when I use the new model file "output_graph.pb", the error shows "Unknown layer type Mean in op InceptionResnetV1/Conv2d_la_3*3/BatchNorm/moments/Mean"

**Comment by dkurt** [comment]
@hhuangfh, try to use `tf.contrib.layers.batch_norm` with the argument `fused=True`. Or fuse constant ops of batch normalization subgraph (see https://github.com/opencv/opencv/issues/9660).

**Comment by hhuangfh** [comment]
@dkurt thanks for your suggestion. I failed at "tensorflow/tools/graph_transforms/transform_graph" because of no package of “ tensorflow.tools.graph_transforms“, or even ”tensorflow.tools". there is no directory like "C:\Anaconda3\Lib\site-packages\tensorflow\tools" or "C:\Anaconda3\Lib\site-packages\tensorflow\tools\graph_transforms". I used anaconda to install tensorflow in win10 64bit.  I tried many ways(like: upgrade tensorflow from  v0.12 to v1.0+, so and so ) to make the command run, as:
`~/tensorflow/bazel-bin/tensorflow/tools/graph_transforms/transform_graph \
  --in_graph=frozen_graph.pb \
  --out_graph=fused_graph.pb \
  --inputs=input_node_name \
  --outputs=output_node_name \
  --transforms="fold_constants sort_by_execution_order"`
, but no one works. As a new player, I have no idea to make it work more. May I ask for more suggestion. Or get a copy of all files under "tensorflow/tools" from you. Many thanks!

**Comment by dkurt** [comment]
@hhuangfh, please follow https://github.com/tensorflow/tensorflow/blob/master/tensorflow/tools/graph_transforms/README.md#using-the-graph-transform-tool.

**Comment by hhuangfh** [comment]
@dkurt Thanks for your guide. I installed a VM of ubuntu16.04, and "bazel build" these tools. the problem "unknown layer type Mean" has been solved. Although there is a new problem "unknown layer type Shape), I think it is a solution to modify parameters of transform_graph's "--transform=". So, maybe I need read the document again. I learned a lot with your help. Thank you!

**Comment by dkurt** [comment]
@hhuangfh, can we close it?

**Comment by frizk123** [comment]
@hhuangfh 你好，opencv读取facenet的问题解决了吗？

**Comment by dkurt** [comment]
@frizk123, please rewrite a question in English to make it easier to answer it.

**Comment by hhuangfh** [comment]
@frizk123  Hi~It's the first time for me to use tensorflow-style model. I got the model from my workmate. And I learned much with the help of @dkurt. 
The model I got is the CKPT style, and I resave it to .pb file to match the requirement of opencv3.3.0. And the next, I use the graph transform tool to process the .pb model. You could remove something which is needless or ohers like avoiding effect of  multi-version. Before using the graph transform tool ,you'd better read the doc  [https://github.com/tensorflow/tensorflow/blob/master/tensorflow/tools/graph_transforms/README.md#using-the-graph-transform-tool](url) first.
If not work, you could use the code `
with tf.gfile.FastGFile('model_file.pb') as f:
    graph_def = tf.GraphDef()
    graph_def.ParseFromString(f.read())
    for node in graph_def.node:
        print(node.op)
`
 to check the types of all nodes. In fact I resaved the .pb file to .pbtxt file and then read the model's text file.
You could read the source code of the project called opencv_dnn in solution of opencv3.3.0.  There are many layers more than twenty at "opencv_dnn/Src/layers".  The simplest way is to debug the program, You could find where is wrong.
All these thing is caused by that I read an article named "OpenCV3.3深度学习模块(DNN)应用-图像分类"
(the application of the module named DNN in opencv3.3-image classification in English) from a WeChat Official Account called "OpenCV学堂"(Opencv School in English), you could read it for details. 
Most of my work is in the architecture of CAFFE. I used many kinds of tasks in CAFFE, and opencv3.3 works well. Because I add some codes into CAFFE, I cut these model to remove needless layers like supervised layers at the phase of application first. 
All these works have been done in a VM of ubuntu16.04, this blog maybe give you a help to install the environment.[http://blog.csdn.net/isuker/article/details/52459992](url)

**Comment by frizk123** [comment]
@hhuangfh Thank you! I try the method you mentioned. However, when I tried to remove some layers which are not supported by opencv (such as Mean), it didn't work. it seems some layers can be removed (Identity, Floor), while others cannot be removed(Switch, Mean) when I used graph transform tool. Do you have any ideas?

**Comment by hhuangfh** [comment]
@frizk123 It won't work absolutely, I think. We couldn't remove hidden nodes(like Mean) from the model, otherwise some nodes in the network  will loss their inputs. And we'll get an unusable model. I don't know much about tensorflow-based networks. But there is a pieces of tensorflow-style model about Mean
'node {
  name: "generator/decoder_7/batchnorm/moments/Mean"
  op: "Mean"
  input: "generator/decoder_7/batchnorm/Identity"
  input: "generator/decoder_7/batchnorm/moments/Mean/reduction_indices"
'
I think this problem has something with Layer BatchNorm. it is possible to use the graph transform tool(GTT) to address this issue. You could ask someone who know very well tensorflow for the detail.
Or, I thought You'd better compiler the source of opencv3.3's newest version, set a breakpoint at class TFImporter::populateNet's implementation and then debug the tensorflow's sample for your model. Check the project, you'll find that opencv_dnn module has been updated.

**Comment by hhuangfh** [comment]
@frizk123 It won't work absolutely, I think. We couldn't remove hidden nodes(like Mean) from the model, otherwise some nodes in the network  will loss their inputs. And we'll get an unusable model. I don't know much about tensorflow-based networks. But there is a pieces of tensorflow-style model about Mean
'node {
  name: "generator/decoder_7/batchnorm/moments/Mean"
  op: "Mean"
  input: "generator/decoder_7/batchnorm/Identity"
  input: "generator/decoder_7/batchnorm/moments/Mean/reduction_indices"
'
I think this problem has something with Layer BatchNorm. it is possible to use the graph transform tool(GTT) to address this issue. You could ask someone who know very well tensorflow for the detail.
Or, I thought You'd better compiler the source of opencv3.3's newest version, set a breakpoint at class TFImporter::populateNet's implementation and then debug the tensorflow's sample for your model. Check the project, you'll find that opencv_dnn module has been updated.
